riscv-sbi/devices/uart-16550.c

```diff
@@ -0,0 +1,122 @@
+// implement uart-16550 driver, only for writing and no interrupts
+
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2019 Western Digital Corporation or its affiliates.
+ *
+ * Authors:
+ *   Anup Patel <anup.patel@wdc.com>
+ */
+
+#include "fdt.h"
+#include "hart.h"
+#include "types.h"
+
+#include "devices/console.h"
+
+
+typedef struct {
+	unsigned int speed;
+	unsigned int divisor;
+	unsigned char mode;
+} baud_t;
+
+
+/* UART 16550 registers */
+enum {
+	rbr = 0, /* Receiver Buffer Register */
+	thr = 0, /* Transmitter Holding Register */
+	dll = 0, /* Divisor Latch LSB */
+	ier = 1, /* Interrupt Enable Register */
+	dlm = 1, /* Divisor Latch MSB */
+	iir = 2, /* Interrupt Identification Register */
+	fcr = 2, /* FIFO Control Register */
+	lcr = 3, /* Line Control Register */
+	mcr = 4, /* Modem Control Register */
+	lsr = 5, /* Line Status Register */
+	msr = 6, /* Modem Status Register */
+	spr = 7, /* Scratch Pad Register */
+};
+
+
+static struct {
+	volatile u8 *base;
+} uart16550_common;
+
+
+static void setReg(u8 reg, u8 val)
+{
+	*(uart16550_common.base + reg) = val;
+	RISCV_FENCE(w, o);
+}
+
+
+static u8 readReg(u8 reg)
+{
+	u8 val = *(uart16550_common.base + reg);
+	RISCV_FENCE(i, r);
```

Comment:
Shouldn't this fence be before reading a value?

Reply:
Both Linux and OpenSBI define MMIO reads with this fence order (and without barrier before read). The read from MMIO will happen before any following reads from main memory which is what we want I guess.

Reply:
Ok, for some reason I was confused and thought that this particular example was in different order than other instances, but it's not.